server/src/main/java/org/opensearch/search/query/QueryPhase.java

```diff
@@ -440,7 +442,29 @@ protected boolean searchWithCollector(
             boolean hasFilterCollector,
             boolean hasTimeout
         ) throws IOException {
-            return QueryPhase.searchWithCollector(searchContext, searcher, query, collectors, hasFilterCollector, hasTimeout);
+            // create the top docs collector last when the other collectors are known
+            final TopDocsCollectorContext topDocsFactory = createTopDocsCollectorContext(searchContext, hasFilterCollector);
+            return searchWithCollector(searchContext, searcher, query, collectors, topDocsFactory, hasFilterCollector, hasTimeout);
+        }
+
+        protected boolean searchWithCollector(
```

Comment:
Why do we need this new method instead of just calling `QueryPhase.searchWithCollector` on L447 above?

Also having 3 `searchWithCollector` implementations that call and override each other is quite confusing to read although I don't have any suggestions for a better name right now